### PR TITLE
Symbol visibility

### PR DESCRIPTION
**Pull request by hartwork**
Regarding the work-in-progress aspect / current CI failure: https://github.com/google/sanitizers/wiki/AddressSanitizerOneDefinitionRuleViolation

@myd7349 @KangLin review and testing very welcome, thank you!

### CODE REVIEW

**Comment by myd7349** [comment]
I see a lot of build errors:
>cmake .. -DURIPARSER_BUILD_DOCS=OFF -DURIPARSER_BUILD_TESTS=OFF

>1>------ Build started: Project: uriparser, Configuration: Debug Win32 ------
1>UriCompare.c
1>uriparser\src\uricompare.c(73): error C2375: 'uriEqualsUriA': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(585): note: see declaration of 'uriEqualsUriA'
1>uriparser\src\uricompare.c(73): error C2375: 'uriEqualsUriW': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(585): note: see declaration of 'uriEqualsUriW'
1>UriEscape.c
1>uriparser\src\uriescape.c(72): error C2375: 'uriEscapeA': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(395): note: see declaration of 'uriEscapeA'
1>uriparser\src\uriescape.c(79): error C2375: 'uriEscapeExA': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(371): note: see declaration of 'uriEscapeExA'
1>uriparser\src\uriescape.c(248): error C2375: 'uriUnescapeInPlaceA': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(440): note: see declaration of 'uriUnescapeInPlaceA'
1>uriparser\src\uriescape.c(254): error C2375: 'uriUnescapeInPlaceExA': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(417): note: see declaration of 'uriUnescapeInPlaceExA'
1>uriparser\src\uriescape.c(72): error C2375: 'uriEscapeW': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(395): note: see declaration of 'uriEscapeW'
1>uriparser\src\uriescape.c(79): error C2375: 'uriEscapeExW': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(371): note: see declaration of 'uriEscapeExW'
1>uriparser\src\uriescape.c(248): error C2375: 'uriUnescapeInPlaceW': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(440): note: see declaration of 'uriUnescapeInPlaceW'
1>uriparser\src\uriescape.c(254): error C2375: 'uriUnescapeInPlaceExW': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(417): note: see declaration of 'uriUnescapeInPlaceExW'
1>UriFile.c
1>uriparser\src\urifile.c(219): error C2375: 'uriUnixFilenameToUriStringA': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(742): note: see declaration of 'uriUnixFilenameToUriStringA'
1>uriparser\src\urifile.c(226): error C2375: 'uriWindowsFilenameToUriStringA': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(765): note: see declaration of 'uriWindowsFilenameToUriStringA'
1>uriparser\src\urifile.c(233): error C2375: 'uriUriStringToUnixFilenameA': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(784): note: see declaration of 'uriUriStringToUnixFilenameA'
1>uriparser\src\urifile.c(240): error C2375: 'uriUriStringToWindowsFilenameA': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(803): note: see declaration of 'uriUriStringToWindowsFilenameA'
1>uriparser\src\urifile.c(219): error C2375: 'uriUnixFilenameToUriStringW': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(742): note: see declaration of 'uriUnixFilenameToUriStringW'
1>uriparser\src\urifile.c(226): error C2375: 'uriWindowsFilenameToUriStringW': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(765): note: see declaration of 'uriWindowsFilenameToUriStringW'
1>uriparser\src\urifile.c(233): error C2375: 'uriUriStringToUnixFilenameW': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(784): note: see declaration of 'uriUriStringToUnixFilenameW'
1>uriparser\src\urifile.c(240): error C2375: 'uriUriStringToWindowsFilenameW': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(803): note: see declaration of 'uriUriStringToWindowsFilenameW'
1>UriIp4.c
1>uriparser\src\uriip4.c(95): error C2375: 'uriParseIpFourAddressA': redefinition; different linkage
1>uriparser\include\uriparser\uriip4.h(98): note: see declaration of 'uriParseIpFourAddressA'
1>uriparser\src\uriip4.c(95): error C2375: 'uriParseIpFourAddressW': redefinition; different linkage
1>uriparser\include\uriparser\uriip4.h(98): note: see declaration of 'uriParseIpFourAddressW'
1>UriMemory.c
1>uriparser\src\urimemory.c(130): error C2375: 'uriEmulateCalloc': redefinition; different linkage
1>uriparser\include\uriparser\uribase.h(318): note: see declaration of 'uriEmulateCalloc'
1>uriparser\src\urimemory.c(153): error C2375: 'uriEmulateReallocarray': redefinition; different linkage
1>uriparser\include\uriparser\uribase.h(338): note: see declaration of 'uriEmulateReallocarray'
1>uriparser\src\urimemory.c(266): error C2375: 'uriCompleteMemoryManager': redefinition; different linkage
1>uriparser\include\uriparser\uribase.h(299): note: see declaration of 'uriCompleteMemoryManager'
1>uriparser\src\urimemory.c(290): error C2375: 'uriTestMemoryManager': redefinition; different linkage
1>uriparser\include\uriparser\uribase.h(363): note: see declaration of 'uriTestMemoryManager'
1>UriNormalize.c
1>uriparser\src\urinormalize.c(478): error C2375: 'uriNormalizeSyntaxMaskRequiredA': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(638): note: see declaration of 'uriNormalizeSyntaxMaskRequiredA'
1>uriparser\src\urinormalize.c(487): error C2375: 'uriNormalizeSyntaxMaskRequiredExA': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(656): note: see declaration of 'uriNormalizeSyntaxMaskRequiredExA'
1>uriparser\src\urinormalize.c(514): error C2375: 'uriNormalizeSyntaxExA': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(678): note: see declaration of 'uriNormalizeSyntaxExA'
1>uriparser\src\urinormalize.c(521): error C2375: 'uriNormalizeSyntaxExMmA': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(700): note: see declaration of 'uriNormalizeSyntaxExMmA'
1>uriparser\src\urinormalize.c(529): error C2375: 'uriNormalizeSyntaxA': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(720): note: see declaration of 'uriNormalizeSyntaxA'
1>uriparser\src\urinormalize.c(478): error C2375: 'uriNormalizeSyntaxMaskRequiredW': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(638): note: see declaration of 'uriNormalizeSyntaxMaskRequiredW'
1>uriparser\src\urinormalize.c(487): error C2375: 'uriNormalizeSyntaxMaskRequiredExW': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(656): note: see declaration of 'uriNormalizeSyntaxMaskRequiredExW'
1>uriparser\src\urinormalize.c(514): error C2375: 'uriNormalizeSyntaxExW': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(678): note: see declaration of 'uriNormalizeSyntaxExW'
1>uriparser\src\urinormalize.c(521): error C2375: 'uriNormalizeSyntaxExMmW': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(700): note: see declaration of 'uriNormalizeSyntaxExMmW'
1>uriparser\src\urinormalize.c(529): error C2375: 'uriNormalizeSyntaxW': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(720): note: see declaration of 'uriNormalizeSyntaxW'
1>UriParse.c
1>uriparser\src\uriparse.c(2157): error C2375: 'uriParseUriExA': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(221): note: see declaration of 'uriParseUriExA'
1>uriparser\src\uriparse.c(2204): error C2375: 'uriParseUriA': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(241): note: see declaration of 'uriParseUriA'
1>uriparser\src\uriparse.c(2214): error C2375: 'uriParseSingleUriA': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(263): note: see declaration of 'uriParseSingleUriA'
1>uriparser\src\uriparse.c(2221): error C2375: 'uriParseSingleUriExA': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(288): note: see declaration of 'uriParseSingleUriExA'
1>uriparser\src\uriparse.c(2232): error C2375: 'uriParseSingleUriExMmA': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(314): note: see declaration of 'uriParseSingleUriExMmA'
1>uriparser\src\uriparse.c(2260): error C2375: 'uriFreeUriMembersA': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(331): note: see declaration of 'uriFreeUriMembersA'
1>uriparser\src\uriparse.c(2266): error C2375: 'uriFreeUriMembersMmA': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(347): note: see declaration of 'uriFreeUriMembersMmA'
1>uriparser\src\uriparse.c(2157): error C2375: 'uriParseUriExW': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(221): note: see declaration of 'uriParseUriExW'
1>uriparser\src\uriparse.c(2204): error C2375: 'uriParseUriW': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(241): note: see declaration of 'uriParseUriW'
1>uriparser\src\uriparse.c(2214): error C2375: 'uriParseSingleUriW': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(263): note: see declaration of 'uriParseSingleUriW'
1>uriparser\src\uriparse.c(2221): error C2375: 'uriParseSingleUriExW': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(288): note: see declaration of 'uriParseSingleUriExW'
1>uriparser\src\uriparse.c(2232): error C2375: 'uriParseSingleUriExMmW': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(314): note: see declaration of 'uriParseSingleUriExMmW'
1>uriparser\src\uriparse.c(2260): error C2375: 'uriFreeUriMembersW': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(331): note: see declaration of 'uriFreeUriMembersW'
1>uriparser\src\uriparse.c(2266): error C2375: 'uriFreeUriMembersMmW': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(347): note: see declaration of 'uriFreeUriMembersMmW'
1>UriQuery.c
1>uriparser\src\uriquery.c(90): error C2375: 'uriComposeQueryCharsRequiredA': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(822): note: see declaration of 'uriComposeQueryCharsRequiredA'
1>uriparser\src\uriquery.c(102): error C2375: 'uriComposeQueryCharsRequiredExA': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(842): note: see declaration of 'uriComposeQueryCharsRequiredExA'
1>uriparser\src\uriquery.c(115): error C2375: 'uriComposeQueryA': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(870): note: see declaration of 'uriComposeQueryA'
1>uriparser\src\uriquery.c(126): error C2375: 'uriComposeQueryExA': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(897): note: see declaration of 'uriComposeQueryExA'
1>uriparser\src\uriquery.c(143): error C2375: 'uriComposeQueryMallocA': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(923): note: see declaration of 'uriComposeQueryMallocA'
1>uriparser\src\uriquery.c(154): error C2375: 'uriComposeQueryMallocExA': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(948): note: see declaration of 'uriComposeQueryMallocExA'
1>uriparser\src\uriquery.c(163): error C2375: 'uriComposeQueryMallocExMmA': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(974): note: see declaration of 'uriComposeQueryMallocExMmA'
1>uriparser\src\uriquery.c(363): error C2375: 'uriFreeQueryListA': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(1063): note: see declaration of 'uriFreeQueryListA'
1>uriparser\src\uriquery.c(369): error C2375: 'uriFreeQueryListMmA': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(1078): note: see declaration of 'uriFreeQueryListMmA'
1>uriparser\src\uriquery.c(384): error C2375: 'uriDissectQueryMallocA': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(999): note: see declaration of 'uriDissectQueryMallocA'
1>uriparser\src\uriquery.c(395): error C2375: 'uriDissectQueryMallocExA': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(1022): note: see declaration of 'uriDissectQueryMallocExA'
1>uriparser\src\uriquery.c(404): error C2375: 'uriDissectQueryMallocExMmA': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(1047): note: see declaration of 'uriDissectQueryMallocExMmA'
1>uriparser\src\uriquery.c(90): error C2375: 'uriComposeQueryCharsRequiredW': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(822): note: see declaration of 'uriComposeQueryCharsRequiredW'
1>uriparser\src\uriquery.c(102): error C2375: 'uriComposeQueryCharsRequiredExW': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(842): note: see declaration of 'uriComposeQueryCharsRequiredExW'
1>uriparser\src\uriquery.c(115): error C2375: 'uriComposeQueryW': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(870): note: see declaration of 'uriComposeQueryW'
1>uriparser\src\uriquery.c(126): error C2375: 'uriComposeQueryExW': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(897): note: see declaration of 'uriComposeQueryExW'
1>uriparser\src\uriquery.c(143): error C2375: 'uriComposeQueryMallocW': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(923): note: see declaration of 'uriComposeQueryMallocW'
1>uriparser\src\uriquery.c(154): error C2375: 'uriComposeQueryMallocExW': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(948): note: see declaration of 'uriComposeQueryMallocExW'
1>uriparser\src\uriquery.c(163): error C2375: 'uriComposeQueryMallocExMmW': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(974): note: see declaration of 'uriComposeQueryMallocExMmW'
1>uriparser\src\uriquery.c(363): error C2375: 'uriFreeQueryListW': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(1063): note: see declaration of 'uriFreeQueryListW'
1>uriparser\src\uriquery.c(369): error C2375: 'uriFreeQueryListMmW': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(1078): note: see declaration of 'uriFreeQueryListMmW'
1>uriparser\src\uriquery.c(384): error C2375: 'uriDissectQueryMallocW': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(999): note: see declaration of 'uriDissectQueryMallocW'
1>uriparser\src\uriquery.c(395): error C2375: 'uriDissectQueryMallocExW': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(1022): note: see declaration of 'uriDissectQueryMallocExW'
1>uriparser\src\uriquery.c(404): error C2375: 'uriDissectQueryMallocExMmW': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(1047): note: see declaration of 'uriDissectQueryMallocExMmW'
1>UriRecompose.c
1>uriparser\src\urirecompose.c(77): error C2375: 'uriToStringCharsRequiredA': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(602): note: see declaration of 'uriToStringCharsRequiredA'
1>uriparser\src\urirecompose.c(85): error C2375: 'uriToStringA': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(620): note: see declaration of 'uriToStringA'
1>uriparser\src\urirecompose.c(77): error C2375: 'uriToStringCharsRequiredW': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(602): note: see declaration of 'uriToStringCharsRequiredW'
1>uriparser\src\urirecompose.c(85): error C2375: 'uriToStringW': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(620): note: see declaration of 'uriToStringW'
1>UriResolve.c
1>uriparser\src\uriresolve.c(298): error C2375: 'uriAddBaseUriA': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(461): note: see declaration of 'uriAddBaseUriA'
1>uriparser\src\uriresolve.c(306): error C2375: 'uriAddBaseUriExA': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(484): note: see declaration of 'uriAddBaseUriExA'
1>uriparser\src\uriresolve.c(314): error C2375: 'uriAddBaseUriExMmA': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(509): note: see declaration of 'uriAddBaseUriExMmA'
1>uriparser\src\uriresolve.c(298): error C2375: 'uriAddBaseUriW': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(461): note: see declaration of 'uriAddBaseUriW'
1>uriparser\src\uriresolve.c(306): error C2375: 'uriAddBaseUriExW': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(484): note: see declaration of 'uriAddBaseUriExW'
1>uriparser\src\uriresolve.c(314): error C2375: 'uriAddBaseUriExMmW': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(509): note: see declaration of 'uriAddBaseUriExMmW'
1>UriShorten.c
1>uriparser\src\urishorten.c(297): error C2375: 'uriRemoveBaseUriA': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(538): note: see declaration of 'uriRemoveBaseUriA'
1>uriparser\src\urishorten.c(307): error C2375: 'uriRemoveBaseUriMmA': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(567): note: see declaration of 'uriRemoveBaseUriMmA'
1>uriparser\src\urishorten.c(297): error C2375: 'uriRemoveBaseUriW': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(538): note: see declaration of 'uriRemoveBaseUriW'
1>uriparser\src\urishorten.c(307): error C2375: 'uriRemoveBaseUriMmW': redefinition; different linkage
1>uriparser\include\uriparser\uri.h(567): note: see declaration of 'uriRemoveBaseUriMmW'
1>Generating Code...
1>Done building project "uriparser.vcxproj" -- FAILED.
========== Build: 0 succeeded, 1 failed, 1 up-to-date, 0 skipped ==========

The declaration of these APIs in header files are decorated by `URI_IMPORT`:
```c
#define URI_IMPORT
```
and the definitions in source files are decorated by `URI_EXPORT`:
```c
# define URI_EXPORT __declspec(dllexport)
```

**Comment by myd7349** [comment]
For MSVC:
```c
#if defined(_MSC_VER)
#  define URI_IMPORT __declspec(dllimport)
#  define URI_EXPORT __declspec(dllexport)
#endif

#if defined(URI_STATIC_LIB)
#  define URI_API
#else
#  if defined(URI_LIBRARY_BUILD)
#    defined URI_API URI_EXPORT
#  else
#    defined URI_API URI_IMPORT
#endif
```
Then we can decorate all public APIs with a single `URI_API` in header files(There is no need to decorate these APIs in the source file).

**Comment by hartwork** [comment]
> and this mismatch causes those errors in the log file.

That's great to know, thank you!

> Then we can decorate all public APIs with a single URI_API in header files(There is no need to decorate these APIs in the source file).

If we only add to headers: Will that not get us into a (different version of the) linking mismatch problem again?

**Comment by myd7349** [comment]
Nope. And this pattern has been taken by many projects: https://github.com/zeromq/libzmq/blob/master/include/zmq.h.

**Comment by hartwork** [comment]
Alright, let's give it a try. Next take coming up in a few minutes.

**Comment by myd7349** [comment]
Maybe I can test it tomorrow. I'm going to sleep now.

**Comment by hartwork** [comment]
:+1: I need to get the CI green first, anyway :smiley: Good night!

File: CMakeLists.txt [hunk 1]
```diff
@@ -143,12 +158,16 @@ set_property(
         PUBLIC_HEADER "${API_HEADER_FILES}"
 )
 
+target_compile_definitions(uriparser PUBLIC -DURI_LIBRARY_BUILD)
```
Comment (myd7349) [inline CMakeLists.txt]: This definition is only useful during the building process of `uriparser`, and should not propagate to the CMake module file `uriparser.cmake`. 
Besides, the `-D` prefix is unnecessary according to: https://cmake.org/cmake/help/latest/command/target_compile_definitions.html

```cmake
target_compile_definitions(uriparser PRIVATE URI_LIBRARY_BUILD)
```
may be better.

File: include/uriparser/UriBase.h [hunk 2]
```diff
@@ -97,6 +97,25 @@
...
+#  define URI_PUBLIC
+# endif
+#else
+# if defined(_MSC_VER)
+#  define URI_PUBLIC __declspec(dllimport)
+# else
+#  define URI_PUBLIC
+# endif
+#endif
+
```
Comment (myd7349) [inline include/uriparser/UriBase.h]: To make static lib build work as well, an additional is needed:
```c
#if defined(URI_STATIC_LIB)
# define URI_PUBLIC
#else
# if defined(URI_LIBRARY_BUILD)
#  if defined(_MSC_VER)
#   define URI_PUBLIC __declspec(dllexport)
#  elif defined(URI_VISIBILITY)
#   define URI_PUBLIC __attribute__ ((visibility("default")))
#  else
#   define URI_PUBLIC
#  endif
# else
#  if defined(_MSC_VER)
#   define URI_PUBLIC __declspec(dllimport)
#  else
#   define URI_PUBLIC
#  endif
# endif
#endif
```
Then, these lines should be added to `CMakeLists.txt`:
```cmake
if (NOT BUILD_SHARED_LIBS)
    target_compile_definitions(uriparser PUBLIC URI_STATIC_LIB)
endif ()
```

Comment (myd7349) [reply include/uriparser/UriBase.h]: If we don't do like this, there will be some link errors:
>3>uriparse.obj : error LNK2019: unresolved external symbol __imp__uriParseUriA referenced in function _main
3>uriparse.obj : error LNK2019: unresolved external symbol __imp__uriFreeUriMembersA referenced in function _main
3>C:\Users\vcpkg\Desktop\uriparser\build\Debug\uriparse.exe : fatal error LNK1120: 2 unresolved externals

File: CMakeLists.txt (hunk 1 above)
Comment (hartwork) [reply CMakeLists.txt]: That's a great catch, thank you!

File: include/uriparser/UriBase.h (hunk 2 above)
Comment (hartwork) [reply include/uriparser/UriBase.h]: I think I have both dynamic and static building fixed now. Thanks a lot for your testing and ideas!

**Comment by hartwork** [comment]
Awesome, merged now. Thanks for your help again!

**Pull request merged**